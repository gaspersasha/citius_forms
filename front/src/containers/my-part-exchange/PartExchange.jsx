import React, { useEffect, useState } from 'react';
import { PropTypes } from 'prop-types';
import cn from 'classnames';
import { endpointFetch, getInfo } from '~utils';
import { Typography, Spinner, Button, Seo } from '~components';
import OffersListItem from './OffersList';

import s from './styles/part-exchange.module.sass';

const PartExchange = ({ cms }) => {
  const [data, setData] = useState([]);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    endpointFetch('getPartExSummary')
      .then((response) => {
        const { error, submittedPartExVehicles } = response;

        if (error) {
          throw error;
        }

        setData(submittedPartExVehicles);
      })
      .finally(() => {
        setIsLoading(false);
      });
  }, []);

  const renderOffersList = data.length
    ? data.map(
        ({
          quoteId,
          tcbgDetailsInGarage: { registration, mileage, valuation },
        }) => (
          <OffersListItem
            key={quoteId}
            quoteId={quoteId}
            registration={registration}
            mileage={mileage}
            valuation={valuation}
          />
        )
      )
    : null;

  if (isLoading) {
    return <Spinner />;
  }

  return (
    <>
      <Seo meta={cms?.meta} />
      <div className={cn(s.root, 'main-width-container')}>
        <Typography type="h1" data-id="page-title" className={s.pageTitle}>
          Your Current Cars
        </Typography>
        <Typography type="h6" className={s.subTitle}>
          Review your online offers below. You can get a new online offer to
          sell your car by clicking the <b>INSTANT OFFER</b> button below.
        </Typography>
        {renderOffersList}
        <div className={s.partxSellBlock}>
          <div className={s.partxSellBlockHead}>
            <Typography className={s.partxSellBlockTittle} type="h3">
              <b>Sell your car now</b>
            </Typography>
          </div>
          <Typography type="h6" className={s.partxSellBlockDescr}>
            Get an instant online offer to sell your car
          </Typography>
          <Button href="/part-exchange" className={s.partxSellBlockButton}>
            Instant offer
          </Button>
        </div>
      </div>
    </>
  );
};

export async function getServerSideProps({ req }) {
  const [url] = req.url.split('?');
  const cms = await getInfo(url);

  return {
    props: {
      cms,
    },
  };
}

PartExchange.propTypes = {
  cms: PropTypes.object.isRequired,
};

export default PartExchange;
