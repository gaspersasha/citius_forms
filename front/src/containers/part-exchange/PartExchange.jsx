import React, { useEffect, useState } from 'react';
import { useRouter } from 'next/router';

import { ProgressBar, Parts, PrivacyNotice, Spinner, Seo } from '~components';
import { fillProgress, scrollTo, endpointFetch } from '~utils';
import { useUserContext } from '~contexts';
import PlateForm from './PlateForm';
import MileageForm from './MileageForm';
import ResultScreen from './ResultScreen';

import s from './styles/part-exchange.module.sass';

const PartExchange = () => {
  const [state, setState] = useState({
    registration: '',
    manufacturer: '',
    model: '',
    colour: '',
    year: '',
    transmission: '',
    engineSize: '',
    fuel: '',
    mileage: '',
    acceptPrivacy: false,
    valuation: '',
    unsubscribe: false,
    condition: '',
    serviceHistory: '',
    step: 0,
  });
  const [formsProgress, setFormsProgress] = useState([
    [1, 0],
    [1, 0],
    [1, 0],
  ]);
  const [isLoading, setIsLoading] = useState(false);

  const {
    pathname,
    query: { productAdvertId, leadgen },
    back,
  } = useRouter();

  useEffect(() => {
    const userDetailsString = sessionStorage.getItem('userDetails');

    const userDetails = userDetailsString && JSON.parse(userDetailsString);

    if (userDetails) {
      sessionStorage.setItem(
        'userDetails',
        JSON.stringify({
          ...userDetails,
          prevRoute: pathname,
        })
      );
    }
  }, [pathname]);

  const { user } = useUserContext();

  const urlParams = { productAdvertId: productAdvertId || '' };

  const deleteEmptyUrlParams = (url) => {
    const payloadParams = { ...url };

    for (const props in payloadParams) {
      if (payloadParams[props].length === 0) {
        delete payloadParams[props];
      }
    }

    return payloadParams;
  };

  const scroll = () => {
    scrollTo('[id="__next"]');
  };

  const setData = (data) => {
    if (!data) return Promise.reject(new Error('Empty object'));
    setState({
      ...state,
      registration: data.registration || state.registration,
      manufacturer: data.manufacturer || state.manufacturer,
      model: data.model || state.model,
      colour: data.colour || state.colour,
      year: data.year || state.year,
      transmission: data.transmission || state.transmission,
      engineSize: data.engineSize || state.engineSize,
      fuel: data.fuel || state.fuel,
      mileage: data.mileage || state.mileage,
      valuation: data.valuation || state.valuation,
      step: data.step || state.step,
      acceptPrivacy: data.acceptPrivacy || state.acceptPrivacy,
      condition: data.condition || state.condition,
      serviceHistory: data.serviceHistory || state.serviceHistory,
    });
  };

  useEffect(() => {
    const isUnsubscribe =
      user.subscribed !== undefined ? !user.subscribed : false;

    setState({ ...state, unsubscribe: isUnsubscribe });

    endpointFetch('startPartEx', urlParams)
      .then((data) => {
        setState({
          ...state,
          details: { ...data.details },
          unsubscribe: !data.details.subscribed || !state.details.subscribed,
        });
        scroll();
      })
      .catch(() => console.log("Can't get user data"));
  }, []);

  const onUnsub = () => setState({ ...state, unsubscribe: !state.unsubscribe });

  const setNextStep = (index) => {
    const [max, curr] = formsProgress[state.step];
    const nextStep = curr + 1;

    setFormsProgress((prevState) => {
      const updateStatuses = [...prevState];

      updateStatuses[index] = [max, nextStep];

      return [...updateStatuses];
    });
  };

  const handleBack = () => {
    const { step } = state;

    setFormsProgress((prevState) => {
      const stepCount = prevState.findIndex(([max, curr]) => curr < max);
      const substep = prevState[stepCount][1];

      // nowhere to go back
      if (stepCount === 0 && substep === 0) return [...prevState];

      const updateStatuses = [...prevState];

      updateStatuses[stepCount][1] = substep - 1;

      return [...updateStatuses];
    });

    if (step > 0) setState({ ...state, step: step - 1 });
    if (step === 0) back();
  };

  const progress = fillProgress(formsProgress);

  return (
    <>
      <Seo />
      <div className="form-container">
        <div className={s.controls}>
          <div className="align-corner">
            <button
              onClick={handleBack}
              className={s.back}
              type="button"
              name="back"
            >
              Back
            </button>
            <span className={s.progress}>{`${progress}% Completed`}</span>
          </div>
          <ProgressBar progress={progress} />
          <Parts part={2} step={state.step + 1} />
        </div>
        <div className="form">
          {isLoading && <Spinner />}
          {!isLoading && state.step === 0 && (
            <PlateForm
              setData={setData}
              payData={state}
              setNextStep={setNextStep}
              productAdvertId={productAdvertId}
              scroll={scroll}
              deleteEmptyUrlParams={deleteEmptyUrlParams}
              urlParams={urlParams}
              leadgen={leadgen}
              setIsLoading={setIsLoading}
            />
          )}

          {!isLoading && state.step === 1 && (
            <MileageForm
              setData={setData}
              payData={state}
              setNextStep={setNextStep}
              scroll={scroll}
              deleteEmptyUrlParams={deleteEmptyUrlParams}
              urlParams={urlParams}
              setIsLoading={setIsLoading}
            />
          )}

          {!isLoading && state.step === 2 && (
            <ResultScreen
              valuation={state.valuation}
              setNextStep={setNextStep}
              productAdvertId={productAdvertId}
              leadgen={leadgen}
              scroll={scroll}
            />
          )}
        </div>
        <PrivacyNotice onChange={onUnsub} unsubscribe={state.unsubscribe} />
      </div>
    </>
  );
};

export default PartExchange;
