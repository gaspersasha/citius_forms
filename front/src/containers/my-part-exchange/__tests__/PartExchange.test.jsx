import React from 'react';
import { act } from 'react-dom/test-utils';
import { endpointFetch } from '~utils';
import PartExchange from '../PartExchange';

jest.mock('~utils', () => ({
  ...jest.requireActual('../../../utils'),
  endpointFetch: jest.fn(),
}));

endpointFetch.mockResolvedValue(global.mockData.myPartex);

const props = {
  cms: {},
};

describe('<PartExchange />', () => {
  let wrapper;
  const getWrapper = (method = shallow) => method(<PartExchange {...props} />);

  it('rendered without errors', async () => {
    wrapper = getWrapper(mount);
    await act(async () => expect(wrapper).toMatchSnapshot());
  });
});
