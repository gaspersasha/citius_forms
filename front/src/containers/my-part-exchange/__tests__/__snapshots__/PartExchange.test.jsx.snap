// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PartExchange /> rendered without errors 1`] = `
<PartExchange
  cms={Object {}}
>
  <Spinner
    className=""
  >
    <div
      className="container"
    >
      <div
        className="loader"
      />
    </div>
  </Spinner>
</PartExchange>
`;
